### PR TITLE
Connect steps using *args

### PR DESCRIPTION
**Pull request by schustmi**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
Briefly describe the changes you have introduced.

### CODE REVIEW

File: src/zenml/steps/base_step.py [hunk 1]
```diff
@@ -291,37 +291,87 @@ def _prepare_parameter_spec(self) -> None:
...
+            raise StepInterfaceError(
+                f"Got unexpected input artifacts for step "
+                f"'{self.step_name}': {unexpected_artifacts}."
+            )
+
+        return combined_artifacts
+
     def __call__(
-        self, **artifacts: BaseArtifact
+        self, *artifacts: Channel, **kw_artifacts: Channel
```
Comment (htahir1) [inline src/zenml/steps/base_step.py:359]: Do we need to expose the tfx-layer here? I liked that it was ZenML BaseArtifact. What would be the motivation for the switch? :-)

File: src/zenml/steps/base_step.py [hunk 2]
```diff
@@ -291,37 +291,87 @@ def _prepare_parameter_spec(self) -> None:
...
+            Dictionary containing both the positional and keyword input
+            artifacts.
+
+        Raises:
+            StepInterfaceError: If there are too many or too few artifacts.
+        """
+        input_artifact_keys = list(self.INPUT_SPEC.keys())
+        if len(artifacts) > len(input_artifact_keys):
+            raise StepInterfaceError(
+                f"Too many input artifacts for step '{self.step_name}'. "
```
Comment (htahir1) [inline src/zenml/steps/base_step.py:315]: Cool one! That was weird in the zenhack

File: src/zenml/steps/base_step.py [hunk 3]
```diff
@@ -291,37 +291,87 @@ def _prepare_parameter_spec(self) -> None:
...
+        for i, artifact in enumerate(artifacts):
+            key = input_artifact_keys[i]
+            combined_artifacts[key] = artifact
+
+        for key, artifact in kw_artifacts.items():
+            if key in combined_artifacts:
+                # an artifact for this key was already set by
+                # the positional input artifacts
+                raise StepInterfaceError(
+                    f"Unexpected keyword argument '{key}' for step "
```
Comment (htahir1) [inline src/zenml/steps/base_step.py:331]: Ok i like this logic

**Review by htahir1 — COMMENTED** [review]
Nice one -- just one comment really

File: src/zenml/steps/base_step.py (hunk 1 above)
Comment (schustmi) [reply src/zenml/steps/base_step.py:359]: I would have liked to keep it a BaseArtifact, but I was debugging and realized it's actually channels. I (wrongly) assumed that these were BaseArtifacts when doing type annotations in the first place I think

File: src/zenml/steps/base_step.py (hunk 2 above)
Comment (schustmi) [reply src/zenml/steps/base_step.py:315]: Yeah that was my motivation to quickly fix this as some of the errors with args/kwargs and strage error messages happened during zenhack

File: src/zenml/steps/base_step.py (hunk 1 above)
Comment (htahir1) [reply src/zenml/steps/base_step.py:359]: But wouldnt this be a bit weird when writing the `connect` function for the user? Or do our BaseStep `__call__` functions all returns Channels?

Comment (schustmi) [reply src/zenml/steps/base_step.py:359]: Yeah they all return channels as well, so it should now be consistent at least

Comment (htahir1) [reply src/zenml/steps/base_step.py:359]: Should we consider returning artifacts?

Comment (schustmi) [reply src/zenml/steps/base_step.py:359]: Would definitely be nicer I think, we'd just have to connect BaseArtifacts to a Channel somehow. Can we add custom properties to a BaseArtifact or does that somehow mess with tfx?

Comment (htahir1) [reply src/zenml/steps/base_step.py:359]: We have to investigate but its linked to the same logic for adding custom properties to the metadata store which keeps popping up. It also is a bit logically messed up as Channels point to artifacts in the TFX design

Comment (schustmi) [reply src/zenml/steps/base_step.py:359]: Yeah that's a little strange then, I added a TODO for now

**Pull request merged**